add Bluetooth and BTLE support

tjobanpu opened the pull request:
Added BTLE and Bluetooth support, this requires kernel to have bluez stack.

This connects #41, connects #90

abhijit-mahajani-imgtec reviewed (changes requested):
you need to add Signed-Off-By please refer https://github.com/CreatorKit/creator-docs/blob/master/ContributorGuide.md

Ham22 commented on target/linux/pistachio/base-files/etc/init.d/bluetooth_init.sh:
```diff
@@ -0,0 +1,41 @@
+# Bring up CSR
+DEVICE=/dev/ttyS0
+BAUD_RATE=115200
+BAUD_RATE_LATER=115200
```
This is not fast enough to stream audio, we can go a lot faster see config from old products.

Ham22 commented on target/linux/pistachio/base-files/etc/init.d/csr:
```diff
@@ -0,0 +1,15 @@
+# Copyright (C) 2007 OpenWrt.org
+
+START=80
+INIT_SCRIPT=/etc/init.d/bluetooth_init.sh
```
Seeing as its a new file can you not just do it the procd way in the first place?

Ham22 commented on target/linux/pistachio/base-files/etc/init.d/csr:
```diff
@@ -0,0 +1,15 @@
+	hciconfig name
+	if [ $? -ne 0 ]
+	then
+		chmod +x $INIT_SCRIPT
```
This should be done by the build system not here

Ham22 commented on target/linux/pistachio/base-files/etc/init.d/bluetooth_init.sh:
```diff
@@ -0,0 +1,41 @@
+// PSKEY_CLOCK_REQUEST_ENABLE
+// Reset to 0x0000 for not using the PIO[2] and PIO[3] and disenabling TXCO
+&0246 = 0000
+EOF
```
Missing WiFi coexistance, MAC address and pretty name PSKs

Ham22 commented on target/linux/pistachio/base-files/etc/init.d/csr:
```diff
@@ -0,0 +1,15 @@
+		chmod +x $INIT_SCRIPT
+		.$INIT_SCRIPT
+	fi
+}
```
We should have a stop function too that toggles the reset line for worst case scenarios. See config from old products.

Ham22 commented on target/linux/pistachio/base-files/etc/init.d/csr:
```diff
@@ -0,0 +1,206 @@
+		fi
+	}
+
+	sleep 5
```
Whats the need for this sleep? Is there not something you can check for rather than blindly waiting?

Ham22 commented on target/linux/pistachio/base-files/etc/init.d/csr:
```diff
@@ -0,0 +1,206 @@
+&2485 = 0009 0001
+
+// PSKEY_coex_transaction_priority_table
+//&2488 = 0001 0001 0001 0001 0001 0001 0001 0001 0001 0001 0001 0000 0001 0000 0000 0000 0001
```
Are these (and others) meant to be commented out?

Ham22 commented on target/linux/pistachio/base-files/etc/init.d/csr:
```diff
@@ -0,0 +1,206 @@
+hciconfig hci0 up &&\
+hciconfig hci0 piscan &&\
+hciconfig hci0 version &&\
+#Grep the last three octets of the BT Address from the "hciconfig" which will be used 
```
There is no use of grep here?

Ham22 commented on target/linux/pistachio/base-files/etc/init.d/csr:
```diff
@@ -0,0 +1,206 @@
+hciconfig hci0 version &&\
+#Grep the last three octets of the BT Address from the "hciconfig" which will be used 
+#for BT Name.
+BT_ADDR=$(hciconfig | awk '/BD Address/{print $3}' | cut -c $BT_ADDR_NOT_VISIBLE-) &&\
```
Would it not be easier to just take the last N chars from the MAC variable you created above?

Ham22 commented on target/linux/pistachio/base-files/etc/init.d/csr:
```diff
@@ -0,0 +1,206 @@
+
+start() {
+	echo -n "Starting Bluetooth..."
+	# TODO: KIN-1986 remove reset_csr here
```
This references an obsolete bug tracker. Also that 'bug' is never going to be fixed sp just remove

Ham22 commented on target/linux/pistachio/marduk/target.mk:
```diff
@@ -13,7 +13,7 @@ DEFAULT_PACKAGES+=kmod-leds-gpio kmod-ledtrig-heartbeat kmod-i2c-core i2c-tools
                   uhttpd uboot-envtools tcpdump board-test proddata \
                   luci kmod-bluetooth kmod-mac802154 \
-                  curl libopenssl ca-certificates
+                  curl libopenssl ca-certificates bluez-utils bluez-libs
```
bluez-utils depends on bluez-libs so need need to add that here

francois-berder-imgtec commented on target/linux/pistachio/base-files/etc/init.d/csr:
```diff
@@ -0,0 +1,206 @@
+bringup_csr() {
+local WIFI_STA_MAC=0x$(hexdump -n 6 -v -e '/1 "%02X"' /proc/device-tree/uccp@18480000/mac-address0)
+local BT_OTP_OFFSET=0x2
+local MAC=$(printf "%x" $((WIFI_STA_MAC + BT_OTP_OFFSET)))
```
The format string should be %012X, just to ensure that the result is 12 characters long and padded by zeros.

francois-berder-imgtec commented on target/linux/pistachio/base-files/etc/init.d/csr:
```diff
@@ -0,0 +1,206 @@
+local WIFI_STA_MAC=0x$(hexdump -n 6 -v -e '/1 "%02X"' /proc/device-tree/uccp@18480000/mac-address0)
+local BT_OTP_OFFSET=0x2
+local MAC=$(printf "%x" $((WIFI_STA_MAC + BT_OTP_OFFSET)))
+MAC=$(echo $BT_MAC | sed 's/.\{2\}/&:/g' | sed 's/.$//')
```
Instead of sed, use macaddr_canonicalize function provided by /lib/functions/system.sh.

francois-berder-imgtec commented on target/linux/pistachio/marduk/target.mk:
```diff
@@ -13,7 +13,7 @@ DEFAULT_PACKAGES+=kmod-leds-gpio kmod-ledtrig-heartbeat kmod-i2c-core i2c-tools
                   uhttpd uboot-envtools tcpdump board-test \
                   luci kmod-bluetooth kmod-mac802154 \
-                  curl libopenssl ca-certificates
+                  curl libopenssl ca-certificates bluez-utils bluez-libs
```
bluez-utils selects bluez-libs. Only bluez-utils are necessary

abhijit-mahajani-imgtec reviewed (changes requested):
No need of mentioning changes as review in the commit message. 
you can remove 
Made the changes in the csr init script as per review

Signed-off-by: Tushar Jobanputra <tushar.jobanputra@imgtec.com>

Why shouldn't we have init script named as bluetooth itself instead of csr?  or at least bt-csr?

Ham22 commented on target/linux/pistachio/base-files/etc/init.d/csr:
```diff
@@ -0,0 +1,206 @@
+source /lib/functions/system.sh
+BT_RST_N_GPIO=0
+{
+grep "Marduk" /proc/device-tree/model
```
What are thoughts on switching this to use the built in openwrt stuff in lib/functions?

abhijit-mahajani-imgtec replied:
We have /lib/pistachio.sh for such functions.

Please refer https://github.com/CreatorDev/openwrt/pull/113/commits/0940117c65a16850540d53318478edfd0bf6dffa

abhijit-mahajani-imgtec commented on target/linux/pistachio/base-files/etc/init.d/csr:
```diff
@@ -0,0 +1,206 @@
+
+
+bringup_csr() {
+local WIFI_STA_MAC=0x$(hexdump -n 6 -v -e '/1 "%02X"' /proc/device-tree/uccp@18480000/mac-address0)
```
i think we should consider moving this into common mac.sh or kind so that all other mac addresses can use it

abhijit-mahajani-imgtec reviewed (changes requested):
Any thoughts on renaming this init script ?

abhijit-mahajani-imgtec commented:
I think we should add tests for bluetooth and ble in board-test, if we are adding support here.

Ham22 commented:
Also agree with renaming the file to just something generic such as bluetooth

tjobanpu commented:
I will rename the init script to bt_csr, and implement the suggestions of Ian and Abhijit and also investigate why the hcitool scan is not providing result

Ham22 commented on target/linux/pistachio/base-files/lib/functions/mac.sh:
```diff
@@ -0,0 +1,28 @@
+# common shell lib to obtain wifi mac address
+
+
+get_wifi_macaddr() {
```
Now that it's refactored out this is very confusing to anyone reading as to why we are grabbing the wifi mac. From a production point of view this has always been known as the PRIMARY_MAC and that name would probably help a lot here in self documenting the code.

Ham22 commented on target/linux/pistachio/base-files/lib/functions/mac.sh:
```diff
@@ -0,0 +1,28 @@
+#!/bin/sh
```
This was refactored out for duplicate code in 6lowpan right? Therefore please add the refactoring for that too.

Ham22 commented on target/linux/pistachio/base-files/etc/init.d/bt-csr:
```diff
@@ -0,0 +1,211 @@
+
+# Configure HCI and get some version info
+FAIL=0
+BT_ADDR_NOT_VISIBLE=10
```
This name confused me, no strong opinion but would something like BT_ADDR_MASK not be more traditional.

Ham22 commented on target/linux/pistachio/base-files/etc/init.d/bt-csr:
```diff
@@ -0,0 +1,211 @@
+	killall bluetoothd
+	# Kill the hciattach process if it exists
+	killall hciattach
+	} > /dev/null 2>&1
```
This entire function can be done in a 1 liner `killall -q bluetoothd hciattach`

Ham22 commented on target/linux/pistachio/base-files/lib/functions/mac.sh:
```diff
@@ -0,0 +1,28 @@
+	local wifi_mac_addr=0x$(hexdump -n 6 -v -e '/1 "%02X"' /proc/device-tree/uccp@18480000/mac-address0);
+	if [ $? -ne 0 ]
+	then
+		echo -e "FAIL" >/dev/null
```
?

Ham22 commented on target/linux/pistachio/base-files/lib/functions/mac.sh:
```diff
@@ -0,0 +1,28 @@
+	if [ $? -ne 0 ]
+	then
+		echo -e "FAIL" >/dev/null
+		exit 1;
```
Using a mix of exit codes and string returns makes an unclear API. Perhaps instead just return empty string and expect that to be tested for by caller?

Whichever way you go with if something goes wrong then their should be a loud WARNING left in logs about it.

Ham22 commented on target/linux/pistachio/base-files/etc/init.d/bt-csr:
```diff
@@ -0,0 +1,211 @@
+hciconfig hci0 up &&\
+hciconfig hci0 piscan &&\
+hciconfig hci0 version &&\
+#search the last three octets of the BT Address from the "hciconfig" which will be used 
```
search?

Ham22 commented on target/linux/pistachio/base-files/etc/init.d/bt-csr:
```diff
@@ -0,0 +1,211 @@
+source /lib/functions/system.sh
+
+BT_RST_N_GPIO=0
+MARDUK_PLATFORM="IMG Marduk - Ci40 with ca8210"
```
Any board with the 'IMG Marduk' prefix would have the same gpio. Could potentially keep support for historical boards with no extra cost?

Ham22 commented on target/linux/pistachio/base-files/etc/init.d/bt-csr:
```diff
@@ -0,0 +1,211 @@
+then
+	BT_RST_N_GPIO=43
+else
+	echo -e "Unkown Platform"
```
Why -e?

abhijit-mahajani-imgtec commented on target/linux/pistachio/base-files/etc/init.d/bt-csr:
```diff
@@ -0,0 +1,211 @@
+BT_RST_N_GPIO=0
+MARDUK_PLATFORM="IMG Marduk - Ci40 with ca8210"
+
+model=$(pistachio_board_model)
```
In fact you can use pistachio_board_name instead of model, since it just need board name not the specific model.

Ham22 commented on target/linux/pistachio/base-files/etc/init.d/bt-csr:
```diff
@@ -0,0 +1,215 @@
+	{
+	# Kill the bluetooth daemon AND hciattach if it exists
+	killall -q bluetoothd hciattach
+	} > /dev/null 2>&1
```
The whole point of -q is none of this redirection is needed anymore and there its so simple this function probably isn't needed either.

Ham22 commented on target/linux/pistachio/base-files/etc/init.d/bt-csr:
```diff
@@ -0,0 +1,215 @@
+}
+
+stop() {
+	echo -n "Stopping Bluetooth:"
```
There will be no output from your takedown function so why a colon and no new line?

Ham22 commented on target/linux/pistachio/base-files/lib/netifd/proto/lowpan.sh:
```diff
@@ -26,11 +27,10 @@ proto_lowpan_setup() {
         if [ ! -e /proc/device-tree/uccp@18480000/mac-address0 ]; then
```
This check shouldn't be needed anymore right?

Instead perhaps do the same as you have done for bluetooth:
if [ -z PRIMARY_MAC_ADDR ] ...

Ham22 commented on target/linux/pistachio/base-files/etc/init.d/bt-csr:
```diff
@@ -0,0 +1,215 @@
+
+board=$(pistachio_board_name)
+
+board_name=${board:0:${#MARDUK_BOARD}}
```
Wasn't the point of changing to board name string when it's always called Marduk anyway so this line isn't needed?

root@OpenWrt:/# cat /tmp/sysinfo/board_name
Marduk

Ham22 reviewed (commented):
Couple of minor cosmetic issues left. Just a bit of thorough testing and this is good to go. Does this need to wait for the wifi coex fix or does it at least work at short range without that?

Ham22 commented on target/linux/pistachio/marduk/target.mk:
```diff
@@ -13,7 +13,7 @@ DEFAULT_PACKAGES+=kmod-leds-gpio kmod-ledtrig-heartbeat kmod-i2c-core i2c-tools
                   kmod-tpm-i2c-infineon kmod-random-tpm tpm-tools \
                   uhttpd uboot-envtools tcpdump board-test \
                   luci kmod-bluetooth kmod-mac802154 \
```
kmod-bluetooth no longer needed explicitly as its pulled in by bluez

Ham22 commented on target/linux/pistachio/marduk/target.mk:
```diff
@@ -13,7 +13,7 @@ DEFAULT_PACKAGES+=kmod-leds-gpio kmod-ledtrig-heartbeat kmod-i2c-core i2c-tools
                   uhttpd uboot-envtools tcpdump board-test \
                   luci kmod-bluetooth kmod-mac802154 \
-                  curl libopenssl ca-certificates
+                  curl libopenssl ca-certificates bluez-utils
```
what about the bluez daemon?

Avinash95 commented on target/linux/pistachio/base-files/lib/netifd/proto/lowpan.sh:
```diff
@@ -26,11 +27,10 @@ proto_lowpan_setup() {
+                # 6LOWPAN_MAC = PRIMARY_MAC_ADDR + 5
                 local LOWPAN_OTP_OFFSET=0x5
-                local LOWPAN_MAC=$(printf "%012X" $((WIFI_STA_MAC + LOWPAN_OTP_OFFSET)))
+                local LOWPAN_MAC=$(printf "%012X" $((PRIMARY_MAC_ADDR + LOWPAN_OTP_OFFSET)))
```
Should we keep the MAC related stuff in mac.sh and let the caller call get_lowpan_mac() ?
This way we can make the this script generic, as we will take the platform specific OTP layout stuff  out from lowpan.sh

Avinash95 commented on target/linux/pistachio/base-files/etc/init.d/bt-csr:
```diff
@@ -0,0 +1,215 @@
+START=80
+STOP=90
+
+source /lib/functions/mac.sh
```
Should we stick to one syntax for including other scripts, as I can see lowpan.sh is using . for it ?

Avinash95 commented on target/linux/pistachio/base-files/etc/init.d/bt-csr:
```diff
@@ -0,0 +1,215 @@
+fi
+
+# Configure HCI and get some version info
+FAIL=0
```
Do we need this variable ? can't we check the return value of the command, as they all are ANDed ?

ghost commented:
Just used this pr to add tests for bluetooth and I found the init script is not loading at boot. It is marked as disabled.

abhijit-mahajani-imgtec commented on target/linux/pistachio/base-files/lib/netifd/proto/lowpan.sh:
```diff
@@ -23,20 +24,18 @@ proto_lowpan_setup() {
-                local WPAN_EUI=$(echo $LOWPAN_MAC | cut -d: -f1,2,3):ff:fe:$(echo $LOWPAN_MAC | cut -d: -f4,5,6)
-                ip link set $interface address $WPAN_EUI
-        fi
+        local PRIMARY_MAC_ADDR=$(get_primary_macaddr)
```
all this business logic of obtaining lowpan mac address from wifi mac (or primary mac) address should be in mac.sh.
Here i was expecting to see just get_lowpan_mac call. and setting up lowpan interface with that mac

abhijit-mahajani-imgtec commented on target/linux/pistachio/base-files/lib/functions/mac.sh:
```diff
@@ -0,0 +1,33 @@
+#!/bin/sh
```
all the mac address related scripts and functions should be moved into different commit as its not part of Bluetooth scripts

abhijit-mahajani-imgtec commented on target/linux/pistachio/base-files/etc/init.d/bt-csr:
```diff
@@ -0,0 +1,201 @@
+
+
+bringup_csr() {
+local PRIMARY_MAC_ADDR=$(get_primary_macaddr)
```
Why don't we move all bluetooth mac address related logic too into mac.sh and just call get_bluetooth_mac function here?

abhijit-mahajani-imgtec reviewed (dismissed):
Do not add "changes as per review" comments in the commit message. Split the mac address related logic into different commit and bluetooth/btle related init scripts into different commit

tjobanpu committed:
pistachio: Add Bluetooth and BTLE support

Add Init script that configures the onboard CSR 8811 Bluetooth chip

tjobanpu committed:
pistachio: Add MAC address get methods for lowpan and BT chips

The Bluetooth and lowpan MAC address will be derived from
the primary MAC address

Ham22 commented:
Needs rebasing

Avinash95 commented on target/linux/pistachio/base-files/lib/functions/mac.sh:
```diff
@@ -0,0 +1,53 @@
+	else
+		local BT_MAC=$(printf "%012x" $((PRIMARY_MAC_ADDR + BT_OTP_OFFSET)))
+	fi
+	BT_MAC=$(macaddr_canonicalize "$BT_MAC")
```
These 2 lines should be in else block ?

Avinash95 commented on target/linux/pistachio/base-files/lib/functions/mac.sh:
```diff
@@ -0,0 +1,53 @@
+
+get_bluetooth_mac()
+{
+	local PRIMARY_MAC_ADDR=$(get_primary_macaddr)
```
get_lowpan_mac and get_blueooth_mac differs significantly although they do the same thing logically

francois-berder-imgtec replied:
@Avinash95 The get_lowpan_mac returns a EUI-64, unlike get_bluetooth_mac.

Avinash95 replied:
@francois-berder-imgtec  I was talking about the way they were implemented, the current implementation for both returns the MAC address

Ham22 commented on target/linux/pistachio/base-files/lib/functions/mac.sh:
```diff
@@ -0,0 +1,53 @@
+
+	if [ -z PRIMARY_MAC_ADDR ]
+	then
+		exit 1
```
The general rule seems to be that these helper functions should never exit as it would kill off the script calling them.

mtusnio commented:
The rc.local hack doesn't work if you use sysupgrade, probably because sysupgrade backs up /etc/ and overrides the file with an older one.

Ham22 commented:
I thought we agreed to put the 'hack' in uci-defaults? There must be a standard for this, worth asking upstream.

abhijit-mahajani-imgtec commented on target/linux/pistachio/base-files/etc/init.d/bt-csr:
```diff
@@ -0,0 +1,191 @@
+#!/bin/sh /etc/rc.common
+#
+# Copyright 2016 by Imagination Technologies Limited and/or its affiliated group companies.
```
Well this is not BSD-3C license as per https://imghub.imgtec.org/display/IM/Ci40+Top+level?preview=%2F71337119%2F72554420%2FCreator+Ci40+3-clause+BSD+license+1Mar2016+RL.docx however in openwrt they need standard OpenWrt license header https://github.com/CreatorDev/openwrt/blob/ci40/target/linux/pistachio/marduk/target.mk

Avinash95 commented on target/linux/pistachio/base-files/lib/netifd/proto/lowpan.sh:
```diff
@@ -22,21 +23,11 @@ proto_lowpan_setup() {
-                local WPAN_EUI=$(echo $LOWPAN_MAC | cut -d: -f1,2,3):ff:fe:$(echo $LOWPAN_MAC | cut -d: -f4,5,6)
-                ip link set $interface address $WPAN_EUI
-        fi
+	local LOWPAN_MAC=$(get_lowpan_mac)
```
previously we calculated EUI only if there is a valid MAC address. In the current implementation, we should check the return value of get_lowpan_mac?

Avinash95 commented on target/linux/pistachio/base-files/etc/init.d/bt-csr:
```diff
@@ -0,0 +1,191 @@
+
+
+bringup_csr() {
+MAC=$(get_bluetooth_mac)
```
indentation ?

Avinash95 commented on target/linux/pistachio/base-files/lib/functions/mac.sh:
```diff
@@ -0,0 +1,51 @@
+# common shell lib to obtain wifi mac address
+
+
+get_primary_macaddr()
```
at some places you are following function_name() { and at some places it is
function_name ()
{

make sure we follow the same guideline

Avinash95 commented on target/linux/pistachio/base-files/etc/init.d/bt-csr:
```diff
@@ -0,0 +1,190 @@
+BT_RST_N_GPIO=0
+MARDUK_BOARD="Marduk"
+
+board=$(pistachio_board_name)
```
#197 is likely to break this, so let's wait until we decide on something?

Ham22 replied:
Disagree this is blocking a release, #197 is not

Avinash95 commented on target/linux/pistachio/base-files/etc/init.d/bt-csr:
```diff
@@ -0,0 +1,190 @@
+		# Set up CSR PSKEYs
+		PSKEYS=/tmp/ci40.psr
+
+	echo "// PSKEY_BDADDR" > $PSKEYS
```
this looks odd, no need of newline

Avinash95 commented on target/linux/pistachio/base-files/etc/init.d/bt-csr:
```diff
@@ -0,0 +1,190 @@
+	fi
+
+# Configure HCI and get some version info
+BT_ADDR_MASK=10
```
indentation ?

Avinash95 replied:
indentation

Avinash95 commented on target/linux/pistachio/base-files/etc/uci-defaults/02_enable_bt:
```diff
@@ -0,0 +1,7 @@
+#!/bin/sh
+ln -s /etc/init.d/bt-csr /etc/rc.d/S22bt-csr
+sleep 2
+/etc/rc.d/S22bt-csr start
```
Won't the scripts from init.d folder will get executed on boot up ?
why do we need to explicitly call it ?

tjobanpu replied:
The scripts from init.d does not get executed at startup but the files in the folder/etc/rc.d get executed, the files in this folder are symbolic links to the corresponding files in the init.d folder, to run the scripts in the init.d after boot we need to create the symbolic file by executing the command  /etc/init.d/bt-csr enable at least once, to avoid doing this manually we are creating the symbolic file in this script and starting it other wise again we will have to wait for the next boot for the script to be triggered automatically at boot

Ham22 replied:
With the fix mentioned above this file is not needed. As the required links get setup by the build system.

Ham22 commented on target/linux/pistachio/base-files/etc/init.d/bt-csr:
```diff
@@ -0,0 +1,189 @@
+START=22
+STOP=22
+
+. /lib/functions/mac.sh
```
Looks like all your issues are because you have code outside of functions. This file is sourced both at build and run time meaning this code runs when it shouldn't and kills things off in both cases.

I suggest hard coding the gpio to 43 for now and moving the sourcing of files to inside the functions that need them.

abhijit-mahajani-imgtec reviewed (commented):
Commit message if too big for "Add MAC address...."

Need to have a line spacing between short commit message and long commit message.

something like

```
pistachio: add get MAC address methods for BT and lowpan

The Bluetooth and lowpan MAC address will be derived from
the primary MAC address

```

abhijit-mahajani-imgtec commented on target/linux/pistachio/base-files/lib/functions/mac.sh:
```diff
@@ -0,0 +1,47 @@
+#!/bin/sh
+#
+# Copyright 2016 by Imagination Technologies Limited and/or its affiliated group companies.
```
Change this license too to OpenWrt GPLV2

abhijit-mahajani-imgtec replied:
license header not changed?

abhijit-mahajani-imgtec commented on target/linux/pistachio/base-files/lib/functions/mac.sh:
```diff
@@ -0,0 +1,47 @@
+		# 6LOWPAN_MAC = PRIMARY_MAC_ADDR + 5
+		local LOWPAN_OTP_OFFSET=0x5
+		local LOWPAN_MAC=$(printf "%012X" $(($PRIMARY_MAC_ADDR + $LOWPAN_OTP_OFFSET)))
+		LOWPAN_MAC=$(macaddr_setbit_la $(macaddr_canonicalize $LOWPAN_MAC))
```
```
root@OpenWrt:/lib/functions# /etc/init.d/bt-csr restart
Stopping Bluetooth
Starting Bluetooth...

root@OpenWrt:/lib/functions# /etc/rc.common: line 48: macaddr_canonicalize: not found

```

abhijit-mahajani-imgtec reviewed (changes requested):
/etc/init.d/bt-csr restart
giving the stack trace

```
root@OpenWrt:/lib/functions# /etc/init.d/bt-csr restart
Stopping Bluetooth
Starting Bluetooth...
root@OpenWrt:/lib/functions# /etc/rc.common: line 48: macaddr_canonicalize: not found
Can't get device info: No such device
Failed to configure
[ 5918.618031] CPU 2 Unable to handle kernel paging request at virtual address 0000002c, epc == 8040b378, ra == 807fc798
[ 5918.629908] Oops[#1]:
[ 5918.632456] CPU: 2 PID: 3028 Comm: kworker/2:2 Not tainted 4.4.14 #3
[ 5918.639591] Workqueue: events 0x8f124810
[ 5918.643973] task: 8f856900 ti: 8fa4e000 task.ti: 8fa4e000
[ 5918.649986] $ 0   : 00000000 80ab0008 00000001 00010000
[ 5918.655830] $ 4   : 0000002c 00000004 1100fc00 ffff00fe
[ 5918.661666] $ 8   : 8fa4ffe0 0000fc00 00000006 00000b21
[ 5918.667508] $12   : 00000000 8f209880 00000000 00000000
[ 5918.673342] $16   : 00000000 00000020 0000002c 00000020
[ 5918.679184] $20   : 0000013c 00000000 80a3e480 8f925000
[ 5918.685017] $24   : 00000000 8045b878
[ 5918.690858] $28   : 8fa4e000 8fa4fdc0 00010000 807fc798
[ 5918.696702] Hi    : 00005595
[ 5918.699908] Lo    : 00000000
[ 5918.703159] epc   : 8040b378 _raw_spin_lock_irqsave+0x10/0x38
[ 5918.709589] ra    : 807fc798 skb_dequeue+0x24/0x88
[ 5918.714923] Status: 1100fc02 KERNEL EXL
[ 5918.719309] Cause : 00800008 (ExcCode 02)
[ 5918.723772] BadVA : 0000002c
[ 5918.726984] PrId  : 0001a120 (MIPS interAptiv (multi))
[ 5918.732704] Modules linked in: uccp420wlan mac80211 cfg80211 compat pppoe ppp_async iptable_nat pppox ppp_generic pistachio nf_nat_ipv4 nf_conntrack_ipv6 nf_conntrack_ipv4 ipt_REJECT ipt_MASQUERADE xt_time xt_tcpudp xt_state xt_nat xt_multiport xt_mark xt_mac xt_limit xt_conntrack xt_comment xt_TCPMSS xt_REDIRECT xt_LOG xt_CT snd_soc_tpa6130a2 slhc sc16is7xx rfcomm pistachio_internal_dac pistachio_event_timer_module nhc_udp nhc_routing nhc_mobility nhc_ipv6 nhc_hop nhc_fragment nhc_dest nf_reject_ipv4 nf_nat_redirect nf_nat_masquerade_ipv4 nf_nat nf_log_ipv4 nf_defrag_ipv6 nf_defrag_ipv4 nf_conntrack_rtcache nf_conntrack iptable_raw iptable_mangle iptable_filter ip_tables img_spdif_out img_spdif_in img_parallel_out img_i2s_out img_i2s_in hidp hci_uart btusb btintel bnep bluetooth ca8210sm mac802154 ieee802154_6lowpan ieee802154_socket ieee802154 6lowpan snd_soc_core i2c_dev ledtrig_heartbeat ip6t_REJECT nf_reject_ipv6 nf_log_ipv6 nf_log_common ip6table_raw ip6table_mangle ip6table_filter ip6_tables x_tables snd_pcm_dmaengine snd_compress snd_pcm_oss snd_mixer_oss snd_pcm snd_timer snd_rawmidi snd_seq_device snd_hwdep snd soundcore ecb cmac leds_gpio aead crypto_null
[ 5918.848254] Process kworker/2:2 (pid: 3028, threadinfo=8fa4e000, task=8f856900, tls=00000000)
[ 5918.857748] Stack : 8fa4fdfc 8f856948 84334f80 00000002 00000000 00000000 8f905bbc 8f126384
          00000008 8f856948 80aa5f80 00000000 84334f80 8f905ba0 00000000 8f905bbc
          80a3e480 0000013c 00000000 8f1248f4 80a3f0c0 00000000 00000008 80456228
          8f856ca8 8f905b80 8f209f00 8f905ba0 84334b40 00000000 84338100 80a40000
          00000000 8f209f18 00000008 80443928 8f209f00 84334b40 84334b40 80985b64
          ...
[ 5918.897449] Call Trace:
[ 5918.900186] [<8040b378>] _raw_spin_lock_irqsave+0x10/0x38
[ 5918.906217] [<807fc798>] skb_dequeue+0x24/0x88
[ 5918.911191] [<8f126384>] h4_recv_buf+0x1108/0x1274 [hci_uart]
[ 5918.917596]
[ 5918.919257]
Code: 30420001  000000c0  3c030001 <c0850000> 00a33021  e0860000  10c0fffc  00053402  30a5ffff
[ 5918.930751] ---[ end trace 1a3e814635b07470 ]---
[ 5918.941149] CPU 2 Unable to handle kernel paging request at virtual address ffffffec, epc == 80449c84, ra == 80444a30
[ 5918.952970] Oops[#2]:
[ 5918.955519] CPU: 2 PID: 3028 Comm: kworker/2:2 Tainted: G      D         4.4.14 #3
[ 5918.963967] task: 8f856900 ti: 8fa4e000 task.ti: 8fa4e000
[ 5918.969979] $ 0   : 00000000 00000001 00000000 84334f80
[ 5918.975824] $ 4   : 8f856900 00000002 00000000 00000006
[ 5918.981657] $ 8   : c6f2f3bf 00000000 ffffffff 0000042f
[ 5918.987499] $12   : 0000001a 00000013 0000000e 00000007
[ 5918.993332] $16   : 00000002 84334f80 00000002 80aa5f80
[ 5918.999173] $20   : 8f856cac 84334f80 80a3f0c0 00000000
[ 5919.005007] $24   : 00000001 8045b878
[ 5919.010849] $28   : 8fa4e000 8fa4fb58 8f856c24 80444a30
[ 5919.016693] Hi    : 00000000
[ 5919.019899] Lo    : 00020c00
[ 5919.023166] epc   : 80449c84 kthread_data+0x4/0xc
[ 5919.028416] ra    : 80444a30 wq_worker_sleeping+0x14/0x100
[ 5919.034524] Status: 1100fc02 KERNEL EXL
[ 5919.038907] Cause : 80800008 (ExcCode 02)
[ 5919.043370] BadVA : ffffffec
[ 5919.046582] PrId  : 0001a120 (MIPS interAptiv (multi))
[ 5919.052302] Modules linked in: uccp420wlan mac80211 cfg80211 compat pppoe ppp_async iptable_nat pppox ppp_generic pistachio nf_nat_ipv4 nf_conntrack_ipv6 nf_conntrack_ipv4 ipt_REJECT ipt_MASQUERADE xt_time xt_tcpudp xt_state xt_nat xt_multiport xt_mark xt_mac xt_limit xt_conntrack xt_comment xt_TCPMSS xt_REDIRECT xt_LOG xt_CT snd_soc_tpa6130a2 slhc sc16is7xx rfcomm pistachio_internal_dac pistachio_event_timer_module nhc_udp nhc_routing nhc_mobility nhc_ipv6 nhc_hop nhc_fragment nhc_dest nf_reject_ipv4 nf_nat_redirect nf_nat_masquerade_ipv4 nf_nat nf_log_ipv4 nf_defrag_ipv6 nf_defrag_ipv4 nf_conntrack_rtcache nf_conntrack iptable_raw iptable_mangle iptable_filter ip_tables img_spdif_out img_spdif_in img_parallel_out img_i2s_out img_i2s_in hidp hci_uart btusb btintel bnep bluetooth ca8210sm mac802154 ieee802154_6lowpan ieee802154_socket ieee802154 6lowpan snd_soc_core i2c_dev ledtrig_heartbeat ip6t_REJECT nf_reject_ipv6 nf_log_ipv6 nf_log_common ip6table_raw ip6table_mangle ip6table_filter ip6_tables x_tables snd_pcm_dmaengine snd_compress snd_pcm_oss snd_mixer_oss snd_pcm snd_timer snd_rawmidi snd_seq_device snd_hwdep snd soundcore ecb cmac leds_gpio aead crypto_null
[ 5919.167853] Process kworker/2:2 (pid: 3028, threadinfo=8fa4e000, task=8f856900, tls=00000000)
[ 5919.177345] Stack : 8f856cac 84334f80 80a3f0c0 00000000 8f856900 80406d64 84334f80 80a30000
          00000001 80a3e480 00000000 804304f4 8f856900 8fa4f898 8f850000 00000038
          00000001 00000000 80a30000 8fa4fbd8 8f856c24 804076fc 00000001 dc8ba60e
          80a30000 8f856900 8fa4f898 80430e44 80af71b0 80a40000 80af0000 8046fec4
          8fa4fbd8 8fa4fbd8 0000000b 80ab0000 8fa4fd08 80981ed8 0000013c 00000000
          ...
[ 5919.217051] Call Trace:
[ 5919.219789] [<80449c84>] kthread_data+0x4/0xc
[ 5919.224649] [<80444a30>] wq_worker_sleeping+0x14/0x100
[ 5919.230390] [<80406d64>] __schedule+0x144/0xa78
[ 5919.235459] [<804076fc>] schedule+0x64/0x7c
[ 5919.240173] [<80430e44>] do_exit+0x914/0x918
[ 5919.244940] [<804141ac>] default_cu2_call+0x0/0x44
[ 5919.250275]
[ 5919.251935]
Code: 0811261a  00452823  8c82037c <03e00008> 8c42ffec  27bdffe0  24060004  afbf001c  8c85037c
```

abhijit-mahajani-imgtec commented on target/linux/pistachio/dts/pistachio/pistachio_marduk_common.dtsi:
```diff
@@ -264,7 +264,7 @@
 &sdhost {
 	status = "okay";
-	bus-width = <1>;
+	bus-width = <4>;
```
why are these dts changes?

abhijit-mahajani-imgtec commented on target/linux/pistachio/base-files/lib/functions/mac.sh:
```diff
@@ -0,0 +1,47 @@
+#
+# common shell lib to obtain wifi mac address
+
+
```
system.sh needs to be sourced in

abhijit-mahajani-imgtec reviewed (changes requested):
Need to have a line spacing between short commit message and long commit message.

abhijit-mahajani-imgtec reviewed (approved):
Verified that hcitool scan shows the devices in bluetooth vicinity and also tested that Creator-Ci40[xx:yy:zz] gets displayed when searched devices from mobile.
